packages/0x.js/src/order_watcher/expiration_watcher.ts

```diff
@@ -1,59 +1,69 @@
                                                  DEFAULT_ORDER_EXPIRATION_CHECKING_INTERVAL_MS;
-        this.expiration = {};
-        const scoreFunction = ((orderHash: string) => {
-            return this.expiration[orderHash].toNumber();
-        }).bind(this);
-        this.orderHashHeapByExpiration = new Heap(scoreFunction);
+        const scoreFunction = (orderHash: string) => this.expiration[orderHash].toNumber();
+        const comparator = (lhs: string, rhs: string) => scoreFunction(lhs) - scoreFunction(rhs);
+        this.orderHashByExpirationRBTree = new RBTree(comparator);
     }
     public subscribe(callback: (orderHash: string) => void): void {
-        this.callbackIfExists = callback;
+        if (!_.isUndefined(this.orderExpirationCheckingIntervalIdIfExists)) {
+            throw new Error(ZeroExError.SubscriptionAlreadyPresent);
+        }
         this.orderExpirationCheckingIntervalIdIfExists = intervalUtils.setAsyncExcludingInterval(
-            this.pruneExpiredOrders.bind(this), this.orderExpirationCheckingIntervalMs,
+            this.pruneExpiredOrders.bind(this, callback), this.orderExpirationCheckingIntervalMs,
         );
     }
     public unsubscribe(): void {
-        intervalUtils.clearAsyncExcludingInterval(this.orderExpirationCheckingIntervalIdIfExists as NodeJS.Timer);
-        delete this.callbackIfExists;
+        if (_.isUndefined(this.orderExpirationCheckingIntervalIdIfExists)) {
+            throw new Error(ZeroExError.SubscriptionNotFound);
+        }
+        intervalUtils.clearAsyncExcludingInterval(this.orderExpirationCheckingIntervalIdIfExists);
+        delete this.orderExpirationCheckingIntervalIdIfExists;
+    }
+    public addOrder(orderHash: string, expirationUnixTimestampMs: BigNumber): void {
+        this.expiration[orderHash] = expirationUnixTimestampMs;
+        this.orderHashByExpirationRBTree.insert(orderHash);
     }
-    public addOrder(orderHash: string, expirationUnixTimestampSec: BigNumber): void {
-        this.expiration[orderHash] = expirationUnixTimestampSec;
-        // We don't remove hashes on order remove because it's slow (linear).
-        // We just skip them later if the order was already removed from the order watcher.
-        this.orderHashHeapByExpiration.push(orderHash);
+    public removeOrder(orderHash: string): void {
+        this.orderHashByExpirationRBTree.remove(orderHash);
+        delete this.expiration[orderHash];
     }
-    private pruneExpiredOrders(): void {
-        const currentUnixTimestampSec = utils.getCurrentUnixTimestamp();
+    private pruneExpiredOrders(callback: (orderHash: string) => void): void {
+        const currentUnixTimestampMs = utils.getCurrentUnixTimestampMs();
         while (
-            this.orderHashHeapByExpiration.size() !== 0 &&
-            this.expiration[this.orderHashHeapByExpiration.head()].lessThan(currentUnixTimestampSec) &&
-            !_.isUndefined(this.callbackIfExists)
```

Comment:
Don't we want to replace this check with `!_.isUndefined(this.orderExpirationCheckingIntervalIdIfExists)`?